language/vm/src/binary_format.rs

```diff
@@ -0,0 +1,320 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+//
+// Revised binary format changes.
+// One of the objective is to create a model for an interpreter to be efficient.
+// We want to also maximize type information we have a call (binding) sites.
+// And we want to be compact.
+//
+
+// The pool of Identifier
+pub type IdentifierPool = Vec<Identifier>;
+// The pool of byte array - Vector<u8> in Move
+pub type U8VectorPool = Vec<Vec<u8>>;
+// The address pool
+pub type AddressPool = Vec<AccountAddress>;
+// the pool of signatures, single pool where every signature is serialized as a single blob.
+// Argument signatures and return signatures are split in 2 blobs (see FunctionHandle)
+pub type Signature = Vec<Type>;
+pub type SignaturePool = Vec<Signature>;
+
+//
+// Scripts and Modules
+//
+
+
+// Scripts and Modules have a common "section" expressed in Rust.
+// The Rust representation may vary... don't know...
+// The split is entirely an abstraction invention because at binary format we only have
+// tables. The order of tables can be anything.
+
+// Common sections
+pub struct CompiledHeader {
+    // handles are references to definitions within and cross modules
+    pub module_handles: Vec<ModuleHandle>,
+    pub struct_handles: Vec<StructHandle>,
+    pub function_handles: Vec<FunctionHandle>,
+    pub field_handles: Vec<FieldHandle>,
+    // specifications are references to a *generic* handle and and its instantiation, whether
+    // partial or complete instantiation
+    pub struct_specs: Vec<StructSpec>,
+    pub function_specs: Vec<FunctionSpec>,
+    pub field_specs: Vec<FieldSpec>,
+    // common blob tables
+    pub signatures: SignaturePool,
+    pub identifiers: IdentifierPool,
+    pub byte_array_pool: U8VectorPool,
+    pub address_pool: AddressPool,
```

Comment:
It seems you forgot to move these into the `DataSegment` and that the definition of `DataSegment` is missing?

Reply:
I did not add the DataSegment yet, I had it but felt it would not help the conversation.
But I plan to talk about it so it's not forgotten